### PR TITLE
[BUG] minirocket: zero division errors #5174

### LINKED ISSUE #5174: [BUG] minirocket: zero division errors
**Describe the bug**

I was having continuously zero div error, I controlled my data if there is any problem, everything seems ok. I created this toy data to pass it did not, I don't understand where the error is. Thank you 

Data I feed is :
```
array([[-1.11679744, -0.51587092, -0.72657659,  0.43865139,  1.07119445,
         1.71726773, -0.20861789, -0.53530849,  0.41952728, -0.37900196,
        -0.0486512 ],
       [ 1.19859572,  0.34663248,  0.27921192, -0.37355076, -0.13697979,
        -1.67660658, -0.10773519, -0.44831427,  0.20088291,  0.52513851,
        -0.27670363],
       [-1.11791871,  1.11150784, -1.90993541, -0.44980476, -1.29726056,
        -0.18258636, -0.16732601,  0.19669475,  0.97499825, -0.63578353,
        -0.77883463],
       [-1.04179935,  1.85160133,  0.62543341,  2.94047445,  0.50961871,
         0.75043038, -1.55057931,  1.20557205, -1.35234062, -0.28655624,
         1.33955653],
       [ 0.34800554, -1.22696909,  1.54467426,  0.66425094, -0.30645109,
        -1.80616433, -0.91447542, -0.98566946, -0.25229076, -0.34297176,
         0.80435544],
       [-0.18199041, -2.76711681, -0.32714412,  0.65552114, -0.45163785,
         0.1802683 ,  0.50184689,  1.10993959,  0.46871781, -0.87338959,
        -1.70479615],
       [-1.16597502,  0.68321007,  1.58746238, -1.59395792, -0.0651295 ,
        -1.41066933, -1.54172287, -0.8631856 ,  0.56209383,  0.00951521,
         1.64238222],
       [ 2.15454663,  0.42041387,  2.10024815, -0.43261881, -0.65729531,
        -0.03767233,  0.05513039, -0.86195179, -1.49870337, -0.2185756 ,
         0.4556254 ],
       [-0.29761231, -0.25637448,  0.57123268,  0.19078971,  0.10959436,
        -0.25712293,  0.44847066,  1.24115329, -0.67280784, -2.30569028,
        -0.54406789],
       [ 0.81654443,  0.28873811,  0.19483554,  0.08360172,  0.15528944,
        -0.93796558,  1.23281942, -0.07048015, -1.12251059,  0.93780015,
         0.53757122],
       [ 2.3510039 ,  1.93766603,  0.91298758,  0.66617998,  0.05425653,
         0.88004109, -1.25916282, -0.33624087,  0.80747314, -0.32694704,
        -1.15154405]])
```

Operation I ran :

```python

transformer = rocket.MiniRocket(num_kernels=5)
# Try to fit the transformer to the data again.
transformer.fit(data)
```

Error:
```
ZeroDivisionError                         Traceback (most recent call last)
Cell In[46], line 3
      1 transformer = rocket.MiniRocket(num_kernels=5)
      2 # Try to fit the transformer to the data again.
----> 3 transformer.fit(data)

File ~\anaconda3\envs\gpu-env-2\lib\site-packages\sktime\transformations\base.py:442, in BaseTransformer.fit(self, X, y)
    439     self._fit(X=X_inner, y=y_inner)
    440 else:
    441     # otherwise we call the vectorized version of fit
--> 442     self._vectorize("fit", X=X_inner, y=y_inner)
    444 # this should happen last: fitted state is set to True
    445 self._is_fitted = True

File ~\anaconda3\envs\gpu-env-2\lib\site-packages\sktime\transformations\base.py:1242, in BaseTransformer._vectorize(self, methodname, **kwargs)
   1239     else:
   1240         transformers_ = self.transformers_
-> 1242     self.transformers_ = X.vectorize_est(
   1243         transformers_, method=methodname, **kwargs
   1244     )
   1245     return self
   1247 if methodname in TRAFO_METHODS:
   1248     # loop through fitted transformers one-by-one, and transform series/panels

File ~\anaconda3\envs\gpu-env-2\lib\site-packages\sktime\datatypes\_vectorize.py:584, in VectorizedDF.vectorize_est(self, estimator, method, args, args_rowvec, return_type, rowname_default, colname_default, varname_of_self, **kwargs)
    581     args_i[varname_of_self] = group
    583 est_i_method = getattr(est_i, method)
--> 584 est_i_result = est_i_method(**args_i)
    586 if group_name is None:
    587     group_name = rowname_default

File ~\anaconda3\envs\gpu-env-2\lib\site-packages\sktime\transformations\base.py:439, in BaseTransformer.fit(self, X, y)
    437 # we call the ordinary _fit if no looping/vectorization needed
    438 if not vectorization_needed:
--> 439     self._fit(X=X_inner, y=y_inner)
    440 else:
    441     # otherwise we call the vectorized version of fit
    442     self._vectorize("fit", X=X_inner, y=y_inner)

File ~\anaconda3\envs\gpu-env-2\lib\site-packages\sktime\transformations\panel\rocket\_minirocket.py:112, in MiniRocket._fit(self, X, y)
    107 if n_timepoints < 9:
    108     raise ValueError(
    109         f"n_timepoints must be >= 9, but found {n_timepoints};"
    110         " zero pad shorter series so that n_timepoints == 9"
    111     )
--> 112 self.parameters = _fit(
    113     X, self.num_kernels, self.max_dilations_per_kernel, random_state
    114 )
    115 return self

File ~\anaconda3\envs\gpu-env-2\lib\site-packages\sktime\transformations\panel\rocket\_minirocket_numba.py:392, in _fit(X, num_features, max_dilations_per_kernel, seed)
    388 _, n_timepoints = X.shape
    390 num_kernels = 84
--> 392 dilations, num_features_per_dilation = _fit_dilations(
    393     n_timepoints, num_features, max_dilations_per_kernel
    394 )
    396 num_features_per_kernel = np.sum(num_features_per_dilation)
    398 quantiles = _quantiles(num_kernels * num_features_per_kernel)

File ~\anaconda3\envs\gpu-env-2\lib\site-packages\sktime\transformations\panel\rocket\_minirocket_numba.py:358, in _fit_dilations(n_timepoints, num_features, max_dilations_per_kernel)
    354 num_features_per_kernel = num_features // num_kernels
    355 true_max_dilations_per_kernel = min(
    356     num_features_per_kernel, max_dilations_per_kernel
    357 )
--> 358 multiplier = num_features_per_kernel / true_max_dilations_per_kernel
    360 max_exponent = np.log2((n_timepoints - 1) / (9 - 1))
    361 dilations, num_features_per_dilation = np.unique(
    362     np.logspace(0, max_exponent, true_max_dilations_per_kernel, base=2).astype(
    363         np.int32
    364     ),
    365     return_counts=True,
    366 )

ZeroDivisionError: division by zero
```

### PR DESCRIPTION
**Pull request by benshaw2**
#### Reference Issues/PRs

Fixes #5174.

#### What does this implement/fix? Explain your changes.

MultiRocket and MiniRocket transforms are not well defined when the parameter num_kernels is less than 84.

The reason at least 84 kernels must be chosen is as follows. Multi/mini rocket transforms rely on precisely 84 "base" kernels. Kernels used to transform the data are generated as dilations of the base 84. When a user specifies the parameter num_kernels, the algorithm generates a number of kernels equal to the nearest multiple of 84 (rounding down), since the number of dilations per kernel is num_kernels // 84. If the parameter num_kernels is less than 84, no transformation kernels can be generated, since the number of dilations per kernel is 0.

The authors did not foresee this issue, since they always set the parameter num_kernels to be in the thousands and recommended this practice to subsequent users.

When the num_kernels is less than 84 for multi/mini rocket transforms, the actual number of kernels used is set to 84. A new attribute called num_kernels_ gives the actual number of kernels to be used. For transforms, it is set during _fit. For classifiers and regressors using the "minirocket" or "multirocket" transforms, it is calculated at initialization.

#### Does your contribution introduce a new dependency? If yes, which one?

No.

#### What should a reviewer concentrate their feedback on?

Does the divide by zero error happen for multirocket or minirocket transforms? Is the num_kernels_ attribute accessible?

#### Did you add any tests for the change?

In _rocket_classifier.py and _rocket_regressor.py, some failing parameter values were added to get_test_params: when the parameter rocket_transform is either "minirocket" or "multirocket" and num_kernels < 84, there should no longer be a division by zero error.

#### Any other comments?

No other comments.

#### PR checklist

Please go through the checklist below. Please feel free to remove points if they are not applicable.

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

### CODE REVIEW

**Review by fkiraly — CHANGES_REQUESTED** [review]
Thanks!

We need to make sure that there is a test that checks that the failure is no longer there.

For this, could you add the failing parameter combinations to `get_test_params`, in each estimator?

**Comment by fkiraly** [comment]
Also, could you kindly explain why the minimum is 84 specifically?
If that is a hard bound, it should be mentioned in the docstrings.

**Comment by benshaw2** [comment]
Thanks @fkiraly for the feedback. I have added some failing parameters to get_test_params in the classifier and regressor python files: is this what is needed? I mentioned in each docstring that num_kernels must be at least 84 for minirocket and multirocket transforms, and I edited the original PR request comment to explain why this must be. I also added my github information to the .all-contributorsrc file. Let me know if anything else needs changing or if anything else should be added.

File: sktime/classification/kernel_based/_rocket_classifier.py [hunk 1]
```diff
@@ -136,6 +138,15 @@ def __init__(
     ):
         self.num_kernels = num_kernels
         self.rocket_transform = rocket_transform
+        
+        if num_kernels < 84:
```
Comment (fkiraly) [inline sktime/classification/kernel_based/_rocket_classifier.py:142]: simpler: `if num_kernels < 84 and rocket_transform in ["multirocket", "minirocket"]`

**Review by fkiraly — CHANGES_REQUESTED** [review]
> The reason at least 84 kernels must be chosen is as follows. Multi/mini rocket transforms rely on precisely 84 "base" kernels. Kernels used to transform the data are generated as dilations of the base 84. When a user specifies the parameter num_kernels, the algorithm generates a number of kernels equal to the nearest multiple of 84 (rounding down), since the number of dilations per kernel is num_kernels // 84. If the parameter num_kernels is less than 84, no transformation kernels can be generated, since the number of dilations per kernel is 0.

Oh, I see. So the "effective number of kernels" is always a multiple of 84.

Then, may I suggest an alternative resolution:

* say in the docstring that number of kernels should be a multiple of 84, otherwise it will be rounded down to the next positive multiple of 84, or up to 84 if lower.
* This way, too small parameters will lead to 84.
* I would also add a `num_kernels_` fitted parameter, written to `self` in `_fit`, which is the actual number of kernels used, so users can check. This should be explained in the `Attributes` section.

This way, passing too small parameters should not cause an error, making the estimator more robust. Of course it needs to be documented that it *always* gets rounded.

File: sktime/classification/kernel_based/_rocket_classifier.py [hunk 2]
```diff
@@ -236,4 +247,4 @@ def get_test_params(cls, parameter_set="default"):
         if parameter_set == "results_comparison":
             return {"num_kernels": 100}
         else:
-            return {"num_kernels": 20}
+            return {"num_kernels": 20, "rocket_transform": "minirocket"}
```
Comment (fkiraly) [inline sktime/classification/kernel_based/_rocket_classifier.py:250]: this should actually be a lits of multiple dictionaries, each being a parameter set that is tested.

ou can also omit the `else` for better readability, as it follows an `if`-`return` statement. If anything past the `if` is reached, we know it is `else` (logically)

**Comment by benshaw2** [comment]
Thanks @fkiraly for the feedback. I have changed the resolution strategy so that no errors are returned at all: when num_kernels is less than 84 for the multirocket or minirocket transform, num_kernels is effectively set to 84. I have also introduced the new num_kernels_ attribute and updated the documentation. For the actual transforms, the num_kernels_ attribute is set during _fit, though for the classifier and regressor using rocket, the attribute is set at initialization. I also believe I addressed your other comments, though once again let me know what other changes need to be made.

File: sktime/classification/kernel_based/_rocket_classifier.py [hunk 3]
```diff
@@ -233,7 +249,8 @@ def get_test_params(cls, parameter_set="default"):
             instance.
             ``create_test_instance`` uses the first (or only) dictionary in ``params``.
         """
-        if parameter_set == "results_comparison":
```
Comment (fkiraly) [inline sktime/classification/kernel_based/_rocket_classifier.py:236]: we should leave the `results_comparison` parameter set here, this is used in consistency experiments.

I.e., there would be an "if", but no "else".

File: sktime/transformations/panel/rocket/_minirocket.py [hunk 4]
```diff
@@ -85,14 +95,14 @@ class MiniRocket(BaseTransformer):
 
     def __init__(
         self,
-        num_kernels=10_000,
+        num_kernels=9_996,
```
Comment (fkiraly) [inline sktime/transformations/panel/rocket/_minirocket.py:98]: we need to leave it at 10.000 as otherwise it may break user code that uses `get_params`.

**Review by fkiraly — CHANGES_REQUESTED** [review]
Great!

I think we can merge this, up to small changes due to the fact that we cannot change standing interfaces (parameter defaults and "comparison" parameter sets), without lengthy deprecation. I would just leave them as they are:

* no change to init defaults
* no removal of the `results_comparison` parameter set

We could also go through the deprecation process https://www.sktime.net/en/stable/developer_guide/deprecation.html if we really want to make the changes, though I feel that's maybe not really worth the effort here.

The overarching reason is "no user code should break without warning if we can avoid it".

**Comment by benshaw2** [comment]
I've now changed the init defaults back and restored the results_comparison parameter set in _rocket_classifier.py. If that's it, great! Otherwise, let me know of any other changes that need to be made. Thanks.

**Review by fkiraly — APPROVED** [review]
That's it & great!

**Pull request merged**